streaming/src/main/scala/io/gearpump/streaming/task/Subscription.scala

```diff
@@ -69,14 +70,18 @@ class Subscription(
     case up: UnicastPartitioner =>
       (msg: Message) => {
         val partition = up.getPartition(msg, parallelism, taskId.index)
-        sendMessage(msg, partition)
+        val msgsSent = sendMessage(msg, partition)
+        List(TaskId(processorId, partition)) -> msgsSent
       }
     case mp: MulticastPartitioner =>
       (msg: Message) => {
         val partitions = mp.getPartitions(msg, parallelism, taskId.index)
-        partitions.map(partition => sendMessage(msg, partition)).sum
+        partitions.foldLeft((List.empty[TaskId], 0)) { (accum, partition) =>
+          val (tasks, msgsSent) = accum
+          val num = sendMessage(msg, partition)
+          (TaskId(processorId, partition) +: tasks) -> (msgsSent + num)
+        }
```

Comment:
In Subscription.scala, if ```life.contains(msg.timestamp)```, then the number of messages were sent is the length of ```partitions```, maybe you could avoid using foldLeft here.